File: src/main/java/org/janelia/saalfeldlab/paintera/serialization/SerializationHelpers.java
package org.janelia.saalfeldlab.paintera.serialization;

import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonSerializationContext;

import java.util.Map;

public class SerializationHelpers {

  public static final String TYPE_KEY = "type";

  public static final String DATA_KEY = "data";

  public static <T> JsonElement serializeWithClassInfo(final T object, final JsonSerializationContext context) {

	return serializeWithClassInfo(object, context, TYPE_KEY, DATA_KEY);
  }

  public static <T> T deserializeFromClassInfo(final JsonObject map, final JsonDeserializationContext context) throws ClassNotFoundException {

	return deserializeFromClassInfo(map, context, TYPE_KEY, DATA_KEY);
  }

  public static <T> JsonElement serializeWithClassInfo(
		  final T object,
		  final JsonSerializationContext context,
		  final String typeKey,
		  final String dataKey) {

	final JsonObject map = new JsonObject();
	map.addProperty(typeKey, object.getClass().getName());
	map.add(dataKey, context.serialize(object));
	return map;
  }

  public static <T> T deserializeFromClassInfo(
		  final JsonObject map,
		  final JsonDeserializationContext context,
		  final String typeKey,
		  final String dataKey) throws ClassNotFoundException {

	String clsName = map.get(typeKey).getAsString();
	clsName = TMP_FQDN_METADATA_MAP.getOrDefault(clsName, clsName);

	final Class<T> clazz = (Class<T>)Class.forName(clsName);
	return context.deserialize(map.get(dataKey), clazz);
  }

  private static final Map<String, String> TMP_FQDN_METADATA_MAP = Map.of(
		  //		  "org.janelia.saalfeldlab.paintera.state.channel.n5.N5BackendChannel", "org.janelia.saalfeldlab.paintera.state.channel.n5.N5MetadataBackendChannel"
  );

}

File: src/main/kotlin/org/janelia/saalfeldlab/paintera/serialization/SerializationHelpers.kt
package org.janelia.saalfeldlab.paintera.serialization

import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.google.gson.JsonSerializationContext
import org.janelia.saalfeldlab.paintera.serialization.GsonExtensions.get

object SerializationHelpers {
	private const val TYPE_KEY = "type"
	private const val DATA_KEY = "data"

	@JvmStatic
	fun <T> serializeWithClassInfo(obj: T, context: JsonSerializationContext): JsonElement {
		return serializeWithClassInfo(obj, context, TYPE_KEY, DATA_KEY)
	}

	@JvmStatic
	@Throws(ClassNotFoundException::class)
	fun <T> deserializeFromClassInfo(map: JsonObject, context: JsonDeserializationContext): T {
		return deserializeFromClassInfo(map, context, TYPE_KEY, DATA_KEY)
	}

	@JvmStatic
	fun <T> serializeWithClassInfo(
		obj: T,
		context: JsonSerializationContext,
		typeKey: String?,
		dataKey: String?
	): JsonElement {
		val map = JsonObject()
		map.addProperty(typeKey, obj!!.javaClass.name)
		map.add(dataKey, context.serialize(obj))
		return map
	}

	@JvmStatic
	@Throws(ClassNotFoundException::class)
	fun <T> deserializeFromClassInfo(
		map: JsonObject,
		context: JsonDeserializationContext,
		typeKey: String?,
		dataKey: String?
	): T {
		val clsName = map[typeKey].asString
		val clazz = Class.forName(clsName) as Class<T>
		return context.deserialize(map[dataKey], clazz)
	}

	fun <T> JsonSerializationContext.withClassInfo(obj: T): JsonElement = serializeWithClassInfo(obj, this)

	inline fun <reified T> JsonDeserializationContext.fromClassInfo(jsonObject: JsonObject?, letIt: (T) -> Unit = {}): T? {
		return jsonObject?.let {
			deserializeFromClassInfo<T>(jsonObject, this)?.also { letIt(it) }
		}
	}

	inline fun <reified T> JsonDeserializationContext.fromClassInfo(json: JsonElement, key: String, letIt: (T) -> Unit = {}): T? {
		return this.fromClassInfo(json[key], letIt)
	}

	inline fun <reified T> JsonDeserializationContext.fromClassInfo(jsonObject: JsonObject?, typeKey: String?, dataKey: String?, letIt: (T) -> Unit = {}): T? {
		return jsonObject?.let {
			deserializeFromClassInfo<T>(jsonObject, this, typeKey, dataKey)?.also { letIt(it) }
		}
	}

	inline fun <reified T> JsonDeserializationContext.fromClassInfo(
		json: JsonElement,
		key: String,
		typeKey: String?,
		dataKey: String?,
		letIt: (T) -> Unit = {}
	): T? {
		return this.fromClassInfo(json[key], typeKey, dataKey, letIt)
	}
}
